### PR TITLE
Ci version updates to resolve deprecated versions causing pipeline failures

### PR DESCRIPTION
**Pull request by dave42w**
## Description

I've gone through the CI warnings and checked versions of things used. Several versions in our CI are either no longer supported or about to be deprecated. I don't know why dependabot hasn't picked up some of these.

I don't know how to test these locally.

It would be great if server.rs could have the rust1.83 warnings fixed. The error didn't make a lot of sense to me (and the ci is now using Rust 1.83 and so getting these clippy warnings for all PR's)

I'll submit a PR for a rewrite of the web_socket test so that doesn't keep failing.

drop python 3.8 everywhere

switch from macos-12 to macos-latest

switch from ubuntu-20.04 to ubuntu-latest
switch from ubuntu-22.04 to ubuntu-latest

switch from uv v3 to v4

switch from python 3.8 to 3.9

switch from codspeed v2 to v3

switch from nox 2024.03.02 to 2024.10.09

switch upload artifact from v3 to v4

switch from download artifact from v3 to v4

This PR fixes https://github.com/sparckles/Robyn/issues/1042

## PR Checklist

Please ensure that:

- [ ] The PR contains a descriptive title
- [ ] The PR contains a descriptive summary of the changes
- [ ] You build and test your changes before submitting a PR.
- [ ] You have added relevant documentation
- [ ] You have added relevant tests. We prefer integration tests wherever possible

## Pre-Commit Instructions:

- [ ] Ensure that you have run the [pre-commit hooks](https://github.com/sparckles/robyn#%EF%B8%8F-to-develop-locally) on your PR.

### CODE REVIEW

**Comment by sansyrox** [comment]
Hey @dave42w 👋 

The CI fails for some reason :/ Could you please have a look?

**Comment by dave42w** [comment]
> Hey @dave42w 👋
> 
> The CI fails for some reason :/ Could you please have a look?

It looks like the aarch64 is not supported on ubuntu-latest as that has switched to ubuntu-24.04 so I'll switch to ubunti-22.04 explicitly for linux-cross builds

File: .github/workflows/codspeed.yml [hunk 1]
```diff
@@ -14,16 +14,16 @@ env:
...
-      
+
       - name: Install uv
-        uses: astral-sh/setup-uv@v3
-      
+        uses: astral-sh/setup-uv@v4
+
       - uses: actions/setup-python@v5
         with:
-          python-version: "3.8"
```
[the remark this answers is not in the document]
Comment (dave42w) [reply .github/workflows/codspeed.yml:26]: Typo, I've fixed with uv should have been version 3 here.

I've just installed the vscode ext for git hub actions. It's complaining about line 36 of codspeed.yaml

```
if: matrix.os == 'macos'
```

saying that "Context access might be invalid: matrix"

I've not touched that? Is the warning incorrect?

**Comment by sansyrox** [comment]
Hey @dave42w 👋 

The ci is still failing ://

> run-on-arch: /home/runner/work/_actions/uraimo/run-on-arch-action/v2/Dockerfiles/Dockerfile.armv7.ubuntu-latest does not exist.

Could you have a look at this?

**Comment by dave42w** [comment]
Ok I'll have a Google 😁

File: .github/workflows/codspeed.yml [hunk 2]
```diff
@@ -14,16 +14,16 @@ env:
 
 jobs:
   benchmarks:
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-latest
```
Comment (sansyrox) [inline .github/workflows/codspeed.yml:17]: why not keep it at 22.04 ?

Comment (dave42w) [reply .github/workflows/codspeed.yml:17]: I was trying to future proof to avoid having to keep coming back to it.

Comment (dave42w) [reply .github/workflows/codspeed.yml]: switched all to Ubuntu-22.04 for consistency

**Comment by dave42w** [comment]
Added commit renaming of .cargo/config to .cargo/cargo.toml to remove a build warning as config is deprecated since cargo 1.38

File: .github/workflows/preview-deployments.yml [hunk 3]
```diff
@@ -75,10 +75,10 @@ jobs:
           cd ~ && python -c 'import robyn'
 
   linux:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04
```
Comment (sansyrox) [inline .github/workflows/preview-deployments.yml:78]: please don't change this.

File: .github/workflows/preview-deployments.yml [hunk 4]
```diff
@@ -101,13 +101,12 @@ jobs:
           cd ~ && python -c 'import robyn'
 
   linux-cross:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04
```
Comment (sansyrox) [inline .github/workflows/preview-deployments.yml:104]: neither this.

File: .cargo/config.toml
Comment (sansyrox) [inline .cargo/config.toml:1]: I think this will cause a lot of crashes. We should not do this in this PR

**Comment by sansyrox** [comment]
Hey @dave42w 👋 


The CI is still failing :// I have some suggestions tho

File: .cargo/config.toml
Comment (dave42w) [reply .cargo/config.toml:1]: The warning is:
```
Warning: `/home/dave/Code/Robyn/Robyn-main/.cargo/config` is deprecated in favor of `config.toml`
note: if you need to support cargo 1.38 or earlier, you can symlink `config` to `config.toml`
```
Would you be ok if I add the symlink? This is from 2019, Is anyone going to be using such an old copy of rust?

Comment (dave42w) [reply .cargo/config.toml:1]: I've taken this put here. Will put in a separate branch

**Comment by dave42w** [comment]
Hi @sansyrox 

Are we getting anywhere with this?

**Comment by sansyrox** [comment]
Hey @dave42w 👋 

Overall, it looks very good to me 😄 I will be merging this once the CI passes 😄

**Comment by sansyrox** [comment]
Hey @dave42w 👋 

> Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/robyn/__init__.py", line 13, in <module>
    from robyn.authentication import AuthenticationHandler
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/robyn/authentication.py", line 4, in <module>
    from robyn.robyn import Headers, Identity, Request, Response
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/robyn/robyn.cpython-312-darwin.so, 0x0002): tried: '/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/robyn/robyn.cpython-312-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/robyn/robyn.cpython-312-darwin.so' (no such file), '/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/robyn/robyn.cpython-312-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64'))


The ci is failing ://

**Comment by dave42w** [comment]
Ah ha,

There are still some `- uses: actions/checkout@v3` which I guess should be `- uses: actions/checkout@v4`
fixing now.

**Comment by dave42w** [comment]
Hi @sansyrox 

I'm hoping that these version changes fix the problems. After all, I haven't changed any of the logic. My changes are only to update the version of macos and ubuntu we use, update the version of the various "uses" and removing python 3.8.

It would be so helpful if we were only getting ci fails for real problems, they get masked when every PR fails every time because of ci issues.

**Comment by sansyrox** [comment]
Hey @dave42w 👋 

The linux builds are still failing 

> n uraimo/run-on-arch-action@v2
Error: run-on-arch: /home/runner/work/_actions/uraimo/run-on-arch-action/v2/Dockerfiles/Dockerfile.aarch64.ubuntu-22.04 does not exist.

**Comment by dave42w** [comment]
Hi @sansyrox 

> uraimo/run-on-arch-action@v2
> Error: run-on-arch: /home/runner/work/_actions/uraimo/run-on-arch-action/v2/Dockerfiles/Dockerfile.aarch64.ubuntu-22.04 does not exist.

Is this in preview-deployments.yml or releasE_CI.yml?

I've googled it. As far as I can tell it is a change in the build process see https://github.com/uraimo/run-on-arch-action/issues/155
```
I think it's something to do with the build process change in https://github.com/docker-library/official-images. See for example https://github.com/docker-library/python/issues/933:

    We've been slowly moving Official Images to a new build process and moved python yesterday, so now the arch-specific namespaces are an image index because they include attached provenance/sbom metadata.
```

There seem to be mixed messages about whether Ubuntu 24.04 or 22.04 will work (but if we use 20.04 we get warnings from GitHub)

1. It might be fixed with Run-On-Arch 2.8.1 so maybe we should temporarily hard code that version on line 126
2. We go back and try earlier Ubuntu versions. Maybe lock the whole of linux-cross to Ubuntu 22.04 first then 20.04 if that doesn't work.

It's probably 10x faster for you to make these changes directly rather than iterate via this pr?

File: .github/workflows/release-CI.yml [hunk 5]
```diff
@@ -70,21 +70,21 @@ jobs:
...
           cd ~ && python -c 'import robyn'
       - name: Upload wheels
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: wheels
           path: dist
   linux:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04
```
Comment (sansyrox) [inline .github/workflows/release-CI.yml:78]: let's just keep this to ubuntu-latest

**Comment by sansyrox** [comment]
@dave42w ,

I think this pr contains to many changes at once. What do you think about leaving ubuntu as the original and changing everything else?

**Comment by dave42w** [comment]
> @dave42w ,
> 
> I think this pr contains to many changes at once. What do you think about leaving ubuntu as the original and changing everything else?

At the moment I think we are going in slow circles.

It's going to take weeks if I make these changes one at a time into PRs as I can't see the effects.

**Comment by sansyrox** [comment]
@dave42w , we should then address it step by step. First drop python3.8, then update uv and codspeed gh actions. and rest of it in a separate PR.

What do you think?

**Comment by dave42w** [comment]
> @dave42w , we should then address it step by step. First drop python3.8, then update uv and codspeed gh actions. and rest of it in a separate PR.
> 
> What do you think?

Now we know what a mess of conflicts there are a clean step by step approach makes sense.

But as only you can apply and check each change it seems to make sense for you to do it as otherwise it takes us both that amount of time.

**Comment by dave42w** [comment]
I'm marking this as closed. I see @sansyrox has started the step by step approach 😁

**Comment by sansyrox** [comment]
Hey @dave42w 👋 

You can take over. I just dropped support for 3.8 as it was blocking another feature implementation. 😅

**Comment by dave42w** [comment]
> Hey @dave42w 👋
> 
> You can take over. I just dropped support for 3.8 as it was blocking another feature implementation. 😅

See https://github.com/sparckles/Robyn/pull/1081 for the first of these. Just updates the versions of the github actions. I figured there might be dependency issues between these so have put all these version updates (and nothing else) into that new PR

**Pull request closed without merging**